tensorzero-core/src/rate_limiting/mod.rs

```diff
@@ -109,19 +111,35 @@ impl RateLimitingConfig {
         scope_info: &'a ScopeInfo<'a>,
         rate_limited_request: &impl RateLimitedRequest,
     ) -> Result<TicketBorrows, Error> {
-        let limits = self.get_active_limits(scope_info);
-        if limits.is_empty() {
-            return Ok(TicketBorrows::empty());
+        // The actual logic of `consume_tickets` is in this nested async block,
+        // so that we can inspect any errors that it produces.
+        let inner = async move {
+            let limits = self.get_active_limits(scope_info);
+            if limits.is_empty() {
+                return Ok(TicketBorrows::empty());
+            }
+            let rate_limit_resource_requests = rate_limited_request.estimated_resource_usage()?;
+            let ticket_requests: Result<Vec<ConsumeTicketsRequest>, Error> = limits
+                .iter()
+                .map(|limit| limit.get_consume_tickets_request(&rate_limit_resource_requests))
+                .collect();
+            let ticket_requests = ticket_requests?;
+            let results = client.consume_tickets(ticket_requests).await?;
+            check_borrowed_rate_limits(&limits, &results)?;
+            TicketBorrows::new(results, limits)
+        };
+        // TODO - decide  how to include the `Vec<ConsumeTicketsRequest>` in the span
+        let span = tracing::info_span!(
+            "consume_tickets",
+            otel.name = "rate_limiting_consume_tickets",
+            success = tracing::field::Empty,
+        );
+        for (key, value) in scope_info.tags {
```

Comment:
this struct will get more complicated as we go on. Can you make this an impl of ScopeInfo so we are less likely to forget to update this as we add stuff?